Generic readers (#376)
metatensor/metatrain: docs/src/advanced-concepts/preparing-generic-targets.rst

Preparing generic targets for reading by metatrain
==================================================

Besides energy-like targets, the library also supports reading (and training on)
more generic targets.

Input file
----------

In order to read a generic target, you will have to specify its layout in the input
file. Suppose you want to learn a target named ``mtt::my_target``, which is
represented as a set of 10 independent per-atom 3D Cartesian vector (we need to
learn 3x10 values for each atom). The ``target`` section in the input file
should look
like this:

.. code-block:: yaml

    targets:
      mtt::my_target:
        read_from: dataset.xyz
        key: my_target
        quantity: ""
        unit: ""
        per_atom: True
        type:
          cartiesian:
            rank: 1
        num_properties: 10

The crucial fields here are:

- ``per_atom``: This field should be set to ``True`` if the target is a per-atom
    property. Otherwise, it should be set to ``False``.
- ``type``: This field specifies the type of the target. In this case, the target is
    a Cartesian vector. The ``rank`` field specifies the rank of the target. For

>     Cartesian vectors, the rank is 1. Other possibilities are ``scalar`` and
>     ``spherical`` (for a spherical tensor).

Reviewer: ```suggestion
    Cartesian vectors, the rank is 1. Other possibilities for the `type` are ``scalar`` or
    ``spherical`` (for a spherical tensor).
```

Preparing generic targets for reading by metatrain
==================================================

Besides energy-like targets, the library also supports reading (and training on)
more generic targets.

Input file
----------

In order to read a generic target, you will have to specify its layout in the input

> file. Suppose you want to learn a target named ``mtt::my_target``, which is a per-atom
> Cartesian vector with 10 properties. The target section in the input file should look

Reviewer 2: ```suggestion
file. Suppose you want to learn a target named ``mtt::my_target``, which is
represented as a set of 10 independent per-atom 3D Cartesian vector (we need to
learn 3x10 values for each atom). The ``target`` section in the input file
should look
```

like this:

.. code-block:: yaml

    targets:
      mtt::my_target:

>         quantity: ""
>         read_from: dataset.xyz
>         key: energy
>         unit: ""

Reviewer 2: ```suggestion
        read_from: dataset.xyz
        key: my_target
        quantity: ""
        unit: ""
```

        per_atom: True
        type:
          cartiesian:
            rank: 1

>         num_properties: 10

Reviewer 2: This could also be called `n_targets`/`num_targets` to clarify that the "properties" are fully independent from one another.

Author: I don't really love `num_targets: 10` here because the total number is "kinda 30"

Reviewer 2: It's 10 independent targets each with 3 vector components. But we can find another name!

Author: I've improved the docs

The crucial fields here are:

- ``per_atom``: This field should be set to ``True`` if the target is a per-atom

>     property. Otherwise, it should be set to ``False``.

Author: Unfortunately this also affects the data structures (not only the readers). IMO it's better to keep it for a future PR, it shouldn't be too complicated

- ``type``: This field specifies the type of the target. In this case, the target is
    a Cartesian vector. The ``rank`` field specifies the rank of the target. For
    Cartesian vectors, the rank is 1. Other possibilities are ``scalar`` and
    ``spherical`` (for a spherical tensor).
- ``num_properties``: This field specifies the number of different properties in the
    target that need to be learned. In this case, there are 10 properties.

A few more words should be spent on ``spherical`` targets. These should be made of a
certain number of irreducible spherical tensors. For example, if you are learning a

> property that can be decomposed into two spherical tensors with L=0 and L=2, the target

Reviewer 2: ```suggestion
property that can be decomposed into two proper spherical tensors with L=0 and L=2, the target
```

section should would look like this:

.. code-block:: yaml

    targets:
      mtt::my_target:
        quantity: ""
        read_from: dataset.xyz
        key: energy
        unit: ""
        per_atom: True
        type:
          spherical:
            irreps:

>                 - o3_lambda: 0
>                   o3_sigma: 1
>                 - o3_lambda: 2
>                   o3_sigma: 1

Reviewer 2: ```suggestion
                - {o3_lambda: 0, o3_sigma: 1}
                - {o3_lambda: 2, o3_sigma: 1}
```

FYI, this is also valid syntax, and might make these things more readable

Author: Very good idea

        num_properties: 10

where ``o3_lambda`` specifies the L value of the spherical tensor and ``o3_sigma`` its
parity with respect to inversion (1 for proper tensors, -1 for pseudo-tensors).

Preparing your targets -- ASE
-----------------------------

If you are using the ASE readers to read your targets, you will have to save them
either in the ``.info`` (if the target is per structure, i.e. not per atom) or in the
``.arrays`` (if the target is per atom) attributes of the ASE atoms object. Then you can
dump the atoms object to a file using ``ase.io.write``.

The ASE reader will automatically try to reshape the target data to the format expected
given the target section in the input file. In case your target data is invalid, an
error will be raised.

Reading targets with more than one spherical tensor is not supported by the ASE reader.
In that case, you should use the metatensor reader.

Preparing your targets -- metatensor
------------------------------------

If you are using the metatensor readers to read your targets, you will have to save them
as a list of ``metatensor.torch.TensorMap`` objects with ``torch.save`` into a file with

> the ``.mts`` extension.

Author: We can modify the name if you want, but it's not a npz for now, it's a `torch.save`d object (pickle?, `.pt` maybe? But I don't want it to have the same extension as exported models. I thought `.mts` was the best solution).

> the ``.mts`` extension.

Reviewer 2: hmm, why use `torch.save` instead of `metatensor.torch.save`?

> the ``.mts`` extension.

Reviewer 2: The second one will be compatible with files saved from metatensor-core

> the ``.mts`` extension.

Author: It can only save single `TensorMap`s, no? We need a list

The metatensor reader will verify that the target data in the input files corresponds to
the metadata in the provided ``TensorMap`` objects. In case of a mismatch, errors will

> be raised.

Author: This is possible but very complicated. `metatrain` must handle cases where there is more than one property (then do you read them? With what names? Do you prepend `mtt::`? Etc. Guessing the number of properties and symmetries isn't easy either because `per_atom` properties in ASE must be flattened)

> be raised.

Reviewer 2: Sorry, bad example. I was thinking of the case where we read from metatensor data, which should already contain all this information, i.e. allowing 

```
mtt::my_target:
        read_from: dataset.npz
        quantity: ""
        unit: ""
```

> be raised.

Author: I don't know if inferring is a good feature to have yet. I would first make sure it works and then we can add it in the future if there is a need for it. This type of dataset is meant for experienced users (for the moment), who can be deliberate about their usage

> be raised.

Reviewer 2: There is no inferring here, the data is already there. It is more about not duplicating information. In the current version, either you copy-paste what's already in your files, or you get an error.

In particular:

- for any target, the ``properties`` of the ``TensorMap`` should have a single name,

>   namely ``properties``, corresponding to ``num_properties`` values, from 0 to

Author: Ok, I will try to lift this constraint
